### PR TITLE
8307425: Socket input stream read burns CPU cycles with back-to-back poll(0) calls

### PR DESCRIPTION
**Pull request by roy-soumadipta**
Backport 73ac710533a45bf5ba17f308aa49556b877b8bf9 to JDK 17u.

Backporting the fix for https://bugs.openjdk.org/browse/JDK-8307425 merged as part of https://github.com/openjdk/jdk/pull/13798#
Almost all the hunks were rejected. DatagramChannelImpl doesn't override park in jdk17, and hence has been left unchanged. Respective hunks for NioSocketImpl and SelChImpl were not cleanly applied due to the difference in how park is implemented for the respective classes between openjdk/jdk17u and openjdk/jdk. In openjdk/jdk in park, current thread is checked to be virtual or not and if that equates to true Poller.poll is used instead of Net.poll.

NioSocketImpl.java in openjdk/jdk17u:
```
private void park(FileDescriptor fd, int event, long nanos) throws IOException {
        long millis;
        if (nanos == 0) {
            millis = -1;
        } else {
            millis = NANOSECONDS.toMillis(nanos);
        }
        Net.poll(fd, event, millis);
    }
```
NioSocketImpl.java in openjdk/jdk:
```
private void park(FileDescriptor fd, int event, long nanos) throws IOException {
        Thread t = Thread.currentThread();
        if (t.isVirtual()) {
            Poller.poll(fdVal(fd), event, nanos, this::isOpen);
            if (t.isInterrupted()) {
                throw new InterruptedIOException();
            }
        } else {
            long millis;
            if (nanos == 0) {
                millis = -1;
            } else {
                millis = NANOSECONDS.toMillis(nanos);
                if (nanos > MILLISECONDS.toNanos(millis)) {
                    // Round up any excess nanos to the nearest millisecond to
                    // avoid parking for less than requested.
                    millis++;
                }
            }
            Net.poll(fd, event, millis);
        }
    }
```

SelChImpl.java in openjdk/jdk17u:
```
default void park(int event, long nanos) throws IOException {
        long millis;
        if (nanos <= 0) {
            millis = -1;
        } else {
            millis = NANOSECONDS.toMillis(nanos);
        }
        Net.poll(getFD(), event, millis);
    }
```
SelChImpl.java in openjdk/jdk17u:
```
default void park(int event, long nanos) throws IOException {
        if (Thread.currentThread().isVirtual()) {
            Poller.poll(getFDVal(), event, nanos, this::isOpen);
        } else {
            long millis;
            if (nanos <= 0) {
                millis = -1;
            } else {
                millis = NANOSECONDS.toMillis(nanos);
                if (nanos > MILLISECONDS.toNanos(millis)) {
                    // Round up any excess nanos to the nearest millisecond to
                    // avoid parking for less than requested.
                    millis++;
                }
            }
            Net.poll(getFD(), event, millis);
        }
    }
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307425](https://bugs.openjdk.org/browse/JDK-8307425): Socket input stream read burns CPU cycles with back-to-back poll(0) calls

### Reviewers
 * [Oli Gillespie](https://openjdk.org/census#ogillespie) (@olivergillespie - no project role)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/1341/head:pull/1341` \
`$ git checkout pull/1341`

Update a local copy of the PR: \
`$ git checkout pull/1341` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/1341/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1341`

View PR using the GUI difftool: \
`$ git pr show -t 1341`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/1341.diff">https://git.openjdk.org/jdk17u-dev/pull/1341.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/1341#issuecomment-1542777066)

### CODE REVIEW

**Review by olivergillespie — APPROVED** [review]
Looks good to me, thanks :)

**Review by shipilev — APPROVED** [review]
Looks fine to me. Please ask @olivergillespie to tag the issue for maintainer approval.

**Comment by RealCLanger** [comment]
> /integrate

You should have waited with /integrate until the backport is approved via JBS label.

**Comment by RealCLanger** [comment]
And also please enable and run GHA for a basic verification of this fix. Thanks.

**Comment by roy-soumadipta** [comment]
> > /integrate
> 
> You should have waited with /integrate until the backport is approved via JBS label.

Thanks, OpenJDK GHA is now running. Workflows were enabled after the PR was raised and didn't run automatically. Apologies for the early /integrate, was not aware of this.

**Comment by RealCLanger** [comment]
Thanks. Backport is approved and somebody can sponsor it now.

**Comment by shipilev** [comment]
@roy-soumadipta, could you please go to GHA checks and re-schedule the GHA run? I'd like to make sure it is completely green. Then I can sponsor.

**Comment by roy-soumadipta** [comment]
Hi @shipilev,

I tried to re-run the failed jobs quite a few times and the failure seems to be unrelated to the change.
```
Prepare all required actions
Getting action download info
Download action repository 'actions/download-artifact@v3' (SHA:9bc31d5ccc31df68ecc42ccf4149144866c47d8a)
Run ./.github/actions/get-bundles
Run actions/download-artifact@v3
Starting download for bundles-linux-x64
Error: Unable to find an artifact with the name: bundles-linux-x64
Run actions/download-artifact@v3
Starting download for bundles-linux-x64
Error: Unable to find an artifact with the name: bundles-linux-x64
```

Is there any suggested course of action for this?

**Comment by shipilev** [comment]
> Is there any suggested course of action for this?

This is probably a GHA infra failure. I see you have another run in cooking, let's see how that one does.

**Comment by roy-soumadipta** [comment]
> @roy-soumadipta, could you please go to GHA checks and re-schedule the GHA run? I'd like to make sure it is completely green. Then I can sponsor.

@shipilev all the checks have passed. The PR is now ready to sponsor.

**Comment by shipilev** [comment]
Good!

/sponsor

**Pull request closed without merging**
